src/core/core/network.cpp

```diff
@@ -240,13 +255,15 @@ namespace dsn
     void rpc_session::send_message(message_ex* msg)
     {
         //dinfo("%s: rpc_id = %016llx, code = %s", __FUNCTION__, msg->header->rpc_id, msg->header->rpc_name);
-        _message_count.fetch_add(1, std::memory_order_relaxed); // -- in unlink_message
 
-        msg->add_ref(); // released in on_send_completed        
+        msg->add_ref(); // released in on_send_completed
+
         uint64_t sig;
         {
             utils::auto_lock<utils::ex_lock_nr> l(_lock);
             msg->dl.insert_before(&_messages);
+            _message_count.fetch_add(1, std::memory_order_relaxed); // -- in unlink_message
```

Comment:
I think it's better to access _message_count all in locks, and not use atomic type.